<div class="container">

    <br><br><br>

    <div class="col-md-12">
        <h1>Resources used by NanoARG</h1>
        <h3>Computational Tools</h3>
        Here is a list of the different programs used by NanoARG for doing the computations:
        <ul>
            <li>
                <strong><a target="_blank" href="https://github.com/bbuchfink/diamond">Diamond</a></strong>: Accelerated BLAST compatible local sequence aligner.
            </li>
            <li>
                <strong><a target="_blank" href="https://bench.cs.vt.edu/deeparg">DeepARG</a></strong>: A deep learning based approach for predicting antibiotic resistance genes from metagenomics data.
            </li>
            <li>
                <strong><a target="_blank" href="https://ccb.jhu.edu/software/centrifuge/manual.shtml">Centrifuge</a></strong>: Is a novel microbial classification engine that enables rapid, accurate, and sensitive labeling of reads and quantification
                of species on desktop computers.
            </li>
            <li>
                <strong> <a target="_blank" href="https://prosite.expasy.org/mydomains/"> ProSite</a></strong>: Online tool for drawing domain arrangements
            </li>
            <li>
                <strong> <a href="http://js.cytoscape.org/" target="_blank"> Cytoscape.js</a></strong>: A javascript library for network visualization.
            </li>
        </ul>

        <h3>Databases</h3>
        <ul>
            <li>
                <strong><a target="_blank" href="https://bench.cs.vt.edu/argminer/#/database">ARGminer/DeepARG</a></strong>: Antibiotic Resistance Database.
            </li>
            <li>
                <strong><a target="_blank" href="https://bench.cs.vt.edu/ftp/data/databases/MGEs90.fasta">NanoARG-MGE database</a></strong>: Mobile Genetic Elements database developed in NanoARG.
            </li>
            <li>
                <strong><a target="_blank" href="http://bacmet.biomedicine.gu.se/">BacMet</a></strong>: Metal Resistance Database.
            </li>
            <li>
                <strong><a target="_blank" href="https://www.uniprot.org/">Functional Genes</a></strong>: UniRef curated protein database for funcional annotation.
            </li>
            <li>
                <strong>Pathogens</strong>: List of pathogens that are critical to human health from the <a href="https://www.who.int/news-room/detail/27-02-2017-who-publishes-list-of-bacteria-for-which-new-antibiotics-are-urgently-needed">World
          Health Organization</a> and the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4871955/">ESKAPE</a> database.
            </li>
        </ul>


        <h3>Tutorials</h3>
        Int his page you will find several useful tutorials for the analysis of nanopore reads after running nanoARG.
        <ul>
            <li>
                <strong> How to visualize relative abundances in R?</strong>: In this tutorial you will learn how to process the output from nanoARG to produce stacked bar visualizations for different analysis (MGEs, MRGs, ARGs)
                <a routerLink="relative_abundance">More</a>
            </li>
            <li>
                <strong> How to process JSON file in Python?</strong>: In this tutorial you will learn how to handle the JSON files generated by nanoARG for each submitted sample. JSON files contains all the results from an individual sample.
                <a routerLink="process_output">More</a>
            </li>
            <li>
                <strong> How to perform error correction using CANU?</strong>: You can apply an error correction agorithm that is very useful for libraries with high coverage depth such as funcional metagenomics. In this tutorial we will show you how
                to perform this task using CANU correct before submitting the data to nanoARG.
                <a routerLink="run_canu">More</a>
            </li>
        </ul>

    </div>