<br>

<div style="text-align:center">
    <p-dataTable #dt [value]="sampleService.samplesByProject" [rows]="5" [paginator]="true" [pageLinks]="3" exportFilename="metadata">
        <!-- <p-header> -->
        <!-- <h4>Samples</h4> -->
        <!-- <div class="ui-helper-clearfix">
                <button class='btn btn-xs btn-default' type="button" label="metadata.csv" (click)="dt.exportCSV()" style="float:left"><i class="fa fa-download"></i> <strong>Metadata</strong></button>
                <button class='btn btn-xs btn-default' type="button" label="type.csv" (click)="download_type('type_csv')" style="float:left"><i class="fa fa-download"></i> <strong>ARG categories</strong></button>
                <button class='btn btn-xs btn-default' type="button" label="type.csv" (click)="download_type('subtype_csv')" style="float:left"><i class="fa fa-download"></i> <strong>ARG Groups</strong></button>
            </div> -->

        <!-- </p-header> -->
        <p-column field='name' header='Sample Name'></p-column>
        <p-column field='primary-group' header='Biome'></p-column>
        <!-- <p-column field='secondary-group' header='Secondary Group'></p-column> -->
        <!-- <p-column field='date' header='Createion Date'></p-column> -->
        <p-column field='status' header='Status'></p-column>

        <p-column header="Actions" styleClass="">
            <template let-sample="rowData" pTemplate="body">
                <a class=" badge bg-red " (click)="removeSample(sample)"   > Remove</a>
                <a class=" badge bg-blue " (click)="rerun(sample)"   >Run</a>
                <a class=" badge bg-black" (click)="view(sample)"   >View</a>
                <!-- <a class="   bg-green" (click)="edit(sample)" pButton icon="fa-pencil" ></a> -->
            </template>
        </p-column>

        <!--<p-column styleClass="col-button">
            <template let-sample="rowData" pTemplate="body">
                <button type="text" (click)="reRun(sample)" pButton icon="fa-play" label="Run"></button>
            </template>
        </p-column>-->

    </p-dataTable>

    <hr>

    <p-growl [value]="msgs"></p-growl>
    <p-confirmDialog header="Confirmation" icon="fa fa-question-circle" width="425"></p-confirmDialog>



</div>




<div class="box box-solid">
    <!-- <div class="box-header"> -->
    <!-- <h3>Results</h3> -->
    <!-- </div> -->
    <div class="box-header">
        <h1> <i class="fa fa-cubes"></i> {{ selected_sample.name }} </h1>
    </div>
    <div class="box-body">

        <div class="row">
            <div class="col-md-12">
                <p>This section shows the primary statistics of the sample <span class="badge bg-blue"> {{ selected_sample.name }} </span> <strong> ( {{selected_sample._id}} ) </strong> including read length distribution of the reads that contain at least
                    one ARG and any other functional genes including: MGEs, MRG. Additionaly, taxonomy annotation is provided for each one of the reads in the sample.</p>
                <br><br>

                <div class="col-md-4 bg-blue">
                    <h3>Read length distribution</h3>
                    <p class="small">This graph shows the distribution of the read length that contains at least one ARG. The length is represented in kilo basepairs. each position represents the number of sequences withing a length 1000kbp range. The data used in this
                        graph corresponds to the reads that contain at least one ARG.</p>
                </div>
                <div class="col-md-8 col-md-offset-0 text-center">
                    <chart class="col-md-12" [options]="line_chart"></chart>
                </div>
            </div>
            <div class="col-md-12">
                <hr>
            </div>
        </div>


        <div class="row">
            <div class="col-md-12">
                <div class="col-md-4 bg-green">
                    <h3>Antibiotic resistance hits distribution</h3>
                    <h4>({{ general_info.ARGs.counts }} Hits)</h4>
                    <p class="small">
                        ARGs are computed using our <strong> newly </strong> developed deepARG: a deep learning approach to predict antibiotic resistance genes in metagenomes <a class="badge bg-black" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0401-z">(Arango-Argoty et., al.)</a>.
                        Because of the high error rate of nanopore reads, the deepARG-LS v1.0 was run with permisive identity parameters (<strong> --iden 30 --cov 80 --evalue 1e-10 --prob 0.8 </strong>) to guarantee the correctness of the classifications.
                    </p>
                </div>
                <div class="col-md-4 text-center">
                    <!-- <h5>Antibiotic Resistance Genes distribution</h5> -->
                    <chart class="col-md-12" [options]="args_distribution_chart"></chart>
                </div>
                <div class="col-md-4 text-center">
                    <!-- <h5>Antibiotic categories distribution</h5> -->
                    <chart class="col-md-12" [options]="antibiotic_distribution_chart"></chart>
                </div>
                <div class="col-md-12">
                    <hr>
                </div>
            </div>
        </div>

        <div class="col-md-12">
            <div class="col-md-4 bg-blue">
                <h3>Distribution of Mobile Genetic Elements (MGEs)</h3>
                <h4>({{ general_info.MGEs.counts }} Hits)</h4>
                <p class="small">
                    Annotations were classfied as mobile genetic elements (MGEs) based on matching metadata from protein sequences to one of the following keywords: transposase, transposon, conjugative, integrase, integron, recombinase, conjugal, mobilization, recombination,
                    plasmid (
                    <a class="badge bg-black" href="https://www.nature.com/articles/nature13377">Forsberg et., al</a>). Protein sequences that matched those keywords were retrieved from the National Center for Biotechnology Information <a class="badge bg-black"
                        href="">NCBI</a>. Then, sequences with an identity greater than 90% were clustered and the centroid sequence was substracted using <a class="badge bg-black" href="">CD-HIT</a>. Nanopore reads are then screened against this customized
                    set of MGEs using diamond with parameters (<strong> --evalue 1e-10 --id 30 --coverage 80 </strong>).
                </p>
            </div>


            <div class="col-md-8 text-center">
                <chart class="col-md-12" [options]="mges_distribution_chart"></chart>
            </div>

            <div class="col-md-12">
                <hr>
            </div>
        </div>
        <div class="col-md-12">
            <div class="col-md-4 bg-black">
                <h3>Distribution of Metal Resistance Genes (MRGs)</h3>
                <h4>({{ general_info.MRGs.counts }} Hits)</h4>
                <p class="small">
                    Metal Resistane Genes (MRG) were downloaded from the <a class="badge bg-black" href="http://bacmet.biomedicine.gu.se/">BacMet v1.0</a> database (<a class="badge bg-black" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3965030/">Pal. et., al</a>)
                    and screened agains the nanopore reads using DIAMOND with parameters <strong> --evalue 1e-10 --id 30 --coverage 80 </strong>.

                </p>
            </div>

            <div class="col-md-6 text-center">
                <chart class="col-md-12" [options]="metal_distribution_chart"></chart>
            </div>
        </div>
    </div>

</div>






<div class="box box-solid">
    <div class="box-header">
        <h3>ARGs, MGEs, MRGs patterns</h3>
        <p class="small">This section shows a linear representation of the nanopore reads with the detected ARGs and neighboring genes. It can help to identify common ARG patterns (the visualization contains only reads with antibiotic resistance genes). <a href=""> UniRef90</a>            was used to screen the nanopore to a broad set of functional genes using <a href="">diamond</a> with parameters <code> --evalue 1e-10 --id 30 --coverage 80 </code>. By using this set of functional genes, it is possible to identify the components
            of each nanopore read. The order of the inner circles indicates: ARGS, MRGs, MGEs, and other functional genes.
        </p>
    </div>
    <div class="box-body">

        <!-- read circular mapping -->
        <div class="col-md-8 col-md-offset-2">
            <div class="text-center" style="border-style: solid; border-width: 1px; border-color:rgba(0,0,0,0.1)">
                <h4> <strong> {{ selected_read.id }} </strong></h4>
                <h5> <strong> (<i>{{ selected_read.taxa_rank }}</i> {{ selected_read.taxa }}) </strong> </h5>
                <!-- <div class="d-inline" id="read_circle_map-1"></div> -->
                <div id="gene_organization" style="width: 90%;"></div>
                <!-- <div class="d-inline" id="read_circle_map-2"></div> -->
            </div>

        </div>

        <div class="col-md-12 text-center">
            <br>
            <p-dataTable #reads_table_vis [value]="reads_table" [rows]="5" [paginator]="true" [pageLinks]="5" exportFilename="reads">
                <p-column field='id' header='Read'>

                    <template let-column_data="rowData" pTemplate="body">
                            <a class="small" (click)="render_read_circular_map(column_data)" > {{ column_data.id }}  </a>
                        </template>

                </p-column>

                <p-column field='len' header='Read Length'></p-column>
                <p-column field='genes' header='Genes'></p-column>
                <p-column field='args' header='ARGs'></p-column>
                <p-column field='mges' header='MGEs'></p-column>
                <p-column field='mrgs' header='MRGs'></p-column>

            </p-dataTable>
        </div>

    </div>
</div>

<!-- <div class="col-md-12 col-md-offset-0"> -->
<div class="">
    <!-- <div class="box-header">
        
    </div> -->

    <div class="box box-solid">
        <div class="row">
            <div class="col-md-12">
                <div class="col-md-12">
                    <h3>Co-occurrence of MGEs, ARGs and MRGs</h3>
                    <p class="">
                        This visualization shows the ARGs, MGEs and MRGs patterns in the sample. Nodes (genes) are connected in the network as long as they are place within the same nanopore read. Node size represent the number of genes, shape represents the different groups
                        (MGEs, ARGs, MRGs), colors represents the different category, e.g., antibiotic resistance genes. The thickness of the edges represent the extent of the pair is found in the sample.
                    </p>
                    <p class="small">
                        Nodes with star shape represent different species. Note that red star nodes represent antibiotic ressistance "priority pathogens" defined by the World Health Organization (WHO) (<a href="http://www.who.int/mediacentre/news/releases/2017/bacteria-antibiotics-needed/en/">see details here</a>)
                        and the ESKAPE pathogens (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4871955/">Santajit., et al</a>)
                    </p>
                </div>
                <div class="col-md-6">
                    <div id="network"></div>
                    <div class="text-center">
                        <span class="small" *ngFor="let item of network_labels">
                        <i [style.color]="item.color" class="fa fa-circle" aria-hidden="true"></i> {{ item.id }}
                    </span>
                    </div>
                </div>
                <div class="col-md-6">
                    <div id="co_occurrence_chords"></div>
                </div>
            </div>
        </div>

    </div>
</div>
<!-- </div> -->



<div class="box box-solid">
    <div class="box-header">
        <h3>Taxonomy Annotation</h3>
        <p class="small">
            The bar charts show the distribution of the <i>species</i> taxonomy rank. Taxonomy distribution is computed using a rapid, accurate and sensitive microbial community classification program called <a href="https://ccb.jhu.edu/software/centrifuge/manual.shtml">centrifuge</a>            <a href="http://genome.cshlp.org/content/early/2016/11/16/gr.210641.116.abstract">(Kim. et., al.)</a>. Reads are screened against the index for Bacteria, Aarchaea, Viruses and Human genomes, using default parameters. The graph displays the
            top 10 <i>species</i> ranked by the number of reads.
        </p>
    </div>
    <div class="box-body">
        <!-- <div class="col-md-6 text-center">
            <chart [options]="taxonomy_sample_chart_genus" style="width:25%"></chart>
        </div> -->
        <div class="ui-helper-clearfix">
            <button class='btn btn-xs btn-default' type="button" label="taxonomy.csv" (click)="taxonomy_table.exportCSV()" style="float:right"><i class="fa fa-download"></i> <strong>download table</strong></button>
        </div>
        <div class="col-md-6">
            <div class="box-body text-center">
                <p-dataTable #taxonomy_table [value]="taxonomy_data" [rows]="10" [paginator]="true" [pageLinks]="5" exportFilename="reads">
                    <p-column field='tax_id' header='Taxa ID'></p-column>
                    <p-column field='tax_rank' header='Taxa Rank'></p-column>
                    <p-column field='name' header='Name'></p-column>
                    <p-column field='num_reads' header='Reads'></p-column>
                    <!-- <p-column field='rpm' header='RPM'></p-column> -->
                    <!-- download link -->
                </p-dataTable>
            </div>
        </div>

        <div class="col-md-6">
            <chart class="col-md-12" [options]="taxonomy_sample_chart_species"></chart>
        </div>



    </div>
</div>


<div class="box box-solid">
    <div class="box-header">
        <h3>Multiple Sample Comparison</h3>
        <p class="small">
            The table below contains the results for all the samples. Including, sample name, gene category, gene name, number of hits in the sample.
        </p>
        <p class="small">
            The relative abundance is computed respect the total copy number of ARGs.
        </p>
    </div>
    <div class="ui-helper-clearfix">
        <button class='btn btn-xs btn-default' type="button" label="data.csv" (click)="reads_table_vis_comparison.exportCSV()" style="float:right"><i class="fa fa-download"></i> <strong>download table</strong></button>
    </div>
    <div class="box-body text-center">
        <p-dataTable #reads_table_vis_comparison [value]="all_samples" [rows]="10" [paginator]="true" [pageLinks]="5" exportFilename="reads">
            <p-column field='sample' header='Sample'></p-column>
            <p-column field='origin' header='Section'></p-column>
            <p-column field='category' header='Category'></p-column>
            <p-column field='id' header='Gene'></p-column>
            <p-column field='size' header='Hits'></p-column>
            <p-column field='rel_abn_args' header='RA #ARGs'></p-column>
            <p-column field='rel_abn_total_reads' header='RA 10K #Reads'></p-column>
            <p-column field='rel_abn_total_genes' header='RA 10K #Genes'></p-column>
            <p-column field='rel_abn_unique_strains' header='RA #Strains'></p-column>

            <!-- download link -->
        </p-dataTable>
    </div>
</div>




<!-- <div class="col-md-12">
    <chart [options]="compare_samles_stacked_bar" style="width:60%;margin: 0 auto"></chart>
</div> -->

<p-growl [(value)]="msgs" baseZIndex="99999" autoZIndex="false"></p-growl>